storage/innobase/log/log0log.cc

```diff
@@ -1135,8 +1135,15 @@ ATTRIBUTE_COLD void logs_empty_and_mark_files_at_shutdown()
 	shutdown, because the InnoDB layer may have committed or
 	prepared transactions and we don't want to lose them. */
 
+	/* if trx_rollback_is_active is fase, rollback thread is
+	either not started yet or has rolled back all active
+	transactions. If that is not a normal shutdown(e.g.
+	error happened), it should not wait for all active
+	transactions to be rolled back.
+	*/
 	if (ulint total_trx = srv_was_started && !srv_read_only_mode
 	    && srv_force_recovery < SRV_FORCE_NO_TRX_UNDO
+	    && trx_rollback_is_active
 	    ? trx_sys.any_active_transactions() : 0) {
```

Comment:
We already have the setting `innodb_fast_shutdown=3` that will cause the rollback of recovered transactions to be aborted when the shutdown is started. The design invariant of the shutdown loop is that there are no concurrent writes by other threads. If you want some additional special handling around this, I think that it would have to be done in `row_undo_step()` or at a higher level by assigning `srv_fast_shutdown=3`.

Reply:
I handled this to fix the failure of tc_heuristic_recover.test. the test will abort the server before binlog recovery. At the time, rollback thread is not  started yet. the test will hang for ever, since there are some active transactions which will never be rolled back.

Reply:
sorry, but I didn't get it. why to assigning srv_fast_shutdown=3 at row_undo_step()?

Reply:
I did not know why this change was added. Now that you mention the test `main.tc_heuristic_recover` as the reason, I see that this test is using fault injection via `debug_dbug="+d,innobase_xa_fail"`. Could this code change be avoided if those parts of the test were changed to also set `innodb_fast_shutdown=3`?

Reply:
the following test without innobase_xa_fail has same behavior. It is designed to stop the server after binlog recovery  if the server starts with --tc-heuristic-recover option.

Reply:
I got your concern, checking trx_rollback_is_active here could cause the server shutdown without waiting for user sessions to finished the transaction.

Reply:
Hi marko, the patch is updated.